design/20210302.certificate-request-policy.md

```diff
@@ -0,0 +1,509 @@
+
+    // +optional
+    AllowedDNSNames *[]string `json:"allowedDNSNames,omitempty"`
+
+    // +optional
+    AllowedIPAddresses *[]string `json:"allowedIPAddresses,omitempty"`
+
+    // +optional
+    AllowedURIs *[]string `json:"allowedURIs,omitempty"`
+
+    // +optional
+    AllowedEmailAddresses *[]string `json:"allowedEmailAddresses,omitempty"`
+
+    // +optional
+    AllowedIssuers *[]cmmeta.ObjectReference `json:"allowedIssuer,omitempty"`
+
+    // +optional
+    AllowedIsCA *bool `json:"allowedIsCA,omitempty"`
+
+    // +optional
+    AllowedUsages *[]cmapi.KeyUsage `json:"allowedUsages,omitempty"`
+
+    // +optional
+    AllowedPrivateKey *PolicyPrivateKey `json:"allowedPrivateKey,omitempty"`
+
+    // +optional
+    ExternalPolicyServers []string `json:"externalPolicyServers,omitempty"`
+}
+
+type PolicyX509Subject struct {
+    // +optional
+    AllowedOrganizations *[]string `json:"allowedOrganizations,omitempty"`
+    // +optional
+    AllowedCountries *[]string `json:"allowedCountries,omitempty"`
+    // +optional
+    AllowedOrganizationalUnits *[]string
+    // +optional
+    AllowedLocalities *[]string `json:"allowedLocalities,omitempty"`
+    // +optional
+    AllowedProvinces *[]string `json:"allowedProvinces,omitempty"`
+    // +optional
+    AllowedStreetAddresses *[]string `json:"allowedStreetAddresses,omitempty"`
+    // +optional
+    AllowedPostalCodes *[]string `json:"allowedPostalCodes,omitempty"`
+    // +optional
+    AllowedSerialNumber *string `json:"allowedSerialNumber,omitempty"`
+}
+
+type CertificateRequestPolicyPrivateKey struct {
+    // +optional
+    AllowedRotationPolicy *cmapi.PrivateKeyRotationPolicy `json:"allowedRotationPolicy,omitempty"`
+    // +optional
+    AllowedEncoding *cmapi.PrivateKeyEncoding `json:"allowedEncoding,omitempty"`
+    // +optional
+    AllowedAlgorithm *cmapi.PrivateKeyAlgorithm `json:"allowedAlgorithm,omitempty"`
+
+    // Values are inclusive (i.e. a min value with 2048 will accept a size of
+    // 2048). MinSize and MaxSize may be the same.
+    // +optional
+    MinSize int `json:"allowedMinSize,omitempty"`
+    // +optional
+    MaxSize int `json:"allowedMaxSize,omitempty"`
+}
+
+type CertificateRequestPolicyStatus struct {
+    // +optional
+	  Conditions []CertificateRequestPolicyCondition `json:"conditions,omitempty"`
+}
+
+type CertificateRequestPolicyCondition struct {
+    Type CertificateRequestPolicyConditionType `json:"type"`
+    Status cmmeta.ConditionStatus `json:"status"`
+
+    // +optional
+    LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+
+    // +optional
+    Reason string `json:"reason,omitempty"`
+
+    // +optional
+    Message string `json:"message,omitempty"`
+
+    ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+type CertificateRequestPolicyConditionType string
+
+const (
+    CertificateRequestPolicyConditionReady CertificateRequestPolicyConditionType = "Ready"
+)
+```
```

Comment:
```go
// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
// +kubebuilder:storageversion

// +k8s:openapi-gen=true
type CertificateRequestPolicy struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec CertificateRequestPolicySpec `json:"spec"`

	// +optional
	Status CertificateRequestPolicyStatus `json:"status"`
}

// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

type CertificateRequestPolicyList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata"`

	Items []CertificateRequestPolicy `json:"items"`
}

type CertificateRequestPolicySpec struct {
	// +optional
	AllowedSubject *PolicyX509Subject `json:"allowedSubject,omitempty"`

	// +optional
	AllowedCommonName *string `json:"subject,omitempty"`

	// Values are inclusive (i.e. a min value with 50s will accept a duration
	// with 50s). MinDuration and MaxDuration may be the same.
	// +optional
	MinDuration *metav1.Duration `json:"minDuration,omitempty"`
	// +optional
	MaxDuration *metav1.Duration `json:"maxDuration,omitempty"`

	// +optional
	AllowedDNSNames *[]string `json:"allowedDNSNames,omitempty"`

	// +optional
	AllowedIPAddresses *[]string `json:"allowedIPAddresses,omitempty"`

	// +optional
	AllowedURIs *[]string `json:"allowedURIs,omitempty"`

	// +optional
	AllowedEmailAddresses *[]string `json:"allowedEmailAddresses,omitempty"`

	// +optional
	AllowedIssuers *[]cmmeta.ObjectReference `json:"allowedIssuer,omitempty"`

	// +optional
	AllowedIsCA *bool `json:"allowedIsCA,omitempty"`

	// +optional
	AllowedUsages *[]cmapi.KeyUsage `json:"allowedUsages,omitempty"`

	// +optional
	AllowedPrivateKey *PolicyPrivateKey `json:"allowedPrivateKey,omitempty"`

	// +optional
	ExternalPolicyServers []string `json:"externalPolicyServers,omitempty"`
}

type PolicyX509Subject struct {
	// +optional
	AllowedOrganizations *[]string `json:"allowedOrganizations,omitempty"`
	// +optional
	AllowedCountries *[]string `json:"allowedCountries,omitempty"`
	// +optional
	AllowedOrganizationalUnits *[]string `json:"allowedOrganizationalUnits,omitempty"`
	// +optional
	AllowedLocalities *[]string `json:"allowedLocalities,omitempty"`
	// +optional
	AllowedProvinces *[]string `json:"allowedProvinces,omitempty"`
	// +optional
	AllowedStreetAddresses *[]string `json:"allowedStreetAddresses,omitempty"`
	// +optional
	AllowedPostalCodes *[]string `json:"allowedPostalCodes,omitempty"`
	// +optional
	AllowedSerialNumber *string `json:"allowedSerialNumber,omitempty"`
}

type PolicyPrivateKey struct {
	// +optional
	AllowedAlgorithm *cmapi.PrivateKeyAlgorithm `json:"allowedAlgorithm,omitempty"`

	// Values are inclusive (i.e. a min value with 2048 will accept a size of
	// 2048). MinSize and MaxSize may be the same.
	// +optional
	MinSize *int `json:"allowedMinSize,omitempty"`
	// +optional
	MaxSize *int `json:"allowedMaxSize,omitempty"`
}

type CertificateRequestPolicyStatus struct {
	// +optional
	Conditions []CertificateRequestPolicyCondition `json:"conditions,omitempty"`
}

type CertificateRequestPolicyCondition struct {
	Type   CertificateRequestPolicyConditionType `json:"type"`
	Status cmmeta.ConditionStatus                `json:"status"`

	// +optional
	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`

	// +optional
	Reason string `json:"reason,omitempty"`

	// +optional
	Message string `json:"message,omitempty"`

	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
}

type CertificateRequestPolicyConditionType string

const (
	CertificateRequestPolicyConditionReady CertificateRequestPolicyConditionType = "Ready"
)
```

Reply:
Would it make sense to show YAML examples instead of (or alongside) the Go API?

YAML feels more readable to "ease" the reading/understanding of the proposed API

Reply:
@JoshVanL

Reply:
Is the yaml examples at the end of the document not sufficient?

Reply:
(I am nitpicking, please feel free to ignore the below comment)

My thinking is that the design doc is meant to be "glanceable" i.e. anyone looking at it for a couple of seconds should be able to understand the API changes. Having a formal definition (= Go code) does not help readability; I think it should appear in some appendix as opposed to in the "core" of the design doc.

The "core" of the design should showcase examples that walk people through the different use-cases and help them understand the need for this new feature. The Go code can be in an appendix that can be referenced to in the "core" design

Reply:
Ah understood. It is my understanding that it's the other way around. `/design` -> devs, `https://cert-manager.io` -> users

Reply:
I was referring to the readers of the design, whoever it may be, not to the users of cert-manager.

My above comment is a mean to say that reading the design doc should not require a ton of brain power. It should be easy to understand its implications. I struggle a lot to follow along and to understand the policy design due to things like abstract Go code (as opposed to reading easy to read YAML examples along with user stories)

Again, I’m OK with Go code (since it also has to be discussed) but it should not be (in my mind) the early focus of designs since you will lose many people that could have been able to give feedback, but they don’t have the brain power/patience to decode what the structs imply.